File: apps/control-api/app/routers/change_execution.py
"""Exact, bounded deployment and audit history for one tenant-owned change."""

import re
from datetime import UTC, datetime
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Response
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy import and_, or_, select
from sqlalchemy.orm import Session

from app.db import get_session
from app.models import AuditEvent, ChangeRequest, Deployment, Environment
from app.security import Identity, ensure_permission, get_identity

router = APIRouter(tags=["policies"])
Digest = str | None


class Wire(BaseModel):
    model_config = ConfigDict(extra="forbid")


class DeploymentHistory(Wire):
    id: str = Field(max_length=64)
    environment_id: str = Field(max_length=64)
    environment_name: str | None = Field(max_length=128)
    binding_id: str | None = Field(max_length=64)
    target: str = Field(max_length=128)
    status: str = Field(max_length=32)
    created_at: datetime
    verification_level: Literal["none", "config_readback", "behavior_enforced", "failed", "stale", "unknown"]
    independent_result: Literal["not_checked", "verified", "mismatch", "unreachable", "no_receipt", "unknown"]
    backend_mutated: bool | None
    error_digest: Digest = Field(pattern=r"^[a-f0-9]{64}$")


class AuditHistory(Wire):
    id: str = Field(max_length=64)
    action: str = Field(max_length=64)
    actor_id: str = Field(max_length=256)
    resource_id: str = Field(max_length=64)
    created_at: datetime
    review_digest: Digest = Field(pattern=r"^[a-f0-9]{64}$")
    error_digest: Digest = Field(pattern=r"^[a-f0-9]{64}$")


class ChangeExecution(Wire):
    schema_version: Literal["change-execution/v1"] = "change-execution/v1"
    change_id: str = Field(max_length=64)
    change_status: str = Field(max_length=32)
    evaluated_at: datetime
    deployments: list[DeploymentHistory] = Field(max_length=100)
    deployments_truncated: bool
    audit_access: Literal["allowed", "denied"]
    audit_events: list[AuditHistory] = Field(max_length=200)
    audit_truncated: bool
    expanded: bool


def _dict(value: object) -> dict:
    return value if isinstance(value, dict) else {}


def _digest(value: object) -> str | None:
    return value if isinstance(value, str) and re.fullmatch(r"[a-f0-9]{64}", value) else None


def _utc(value: datetime) -> datetime:
    # DB DateTime columns historically store naive UTC.
    return value.replace(tzinfo=UTC) if value.tzinfo is None else value.astimezone(UTC)


def _deployment(row: Deployment, names: dict[str, str]) -> DeploymentHistory:
    verification = _dict(row.verification)
    level = verification.get("level", verification.get("verification_level"))
    aliases = {
        "readback_verified": "config_readback",
        "config_readback": "config_readback",
        "enforcement_verified": "behavior_enforced",
        "behavior_verified": "behavior_enforced",
        "failed": "failed",
        "error": "failed",
        "stale": "stale",
        "expired": "stale",
    }
    projected_level = aliases.get(level, "unknown") if isinstance(level, str) and level else "none"
    if verification.get("stale") is True or verification.get("expired") is True:
        projected_level = "stale"
    attestation = _dict(verification.get("independent_attestation"))
    result = attestation.get("result")
    independent = (
        result
        if result in ("verified", "mismatch", "unreachable", "no_receipt")
        else "unknown"
        if attestation
        else "not_checked"
    )
    changed = verification.get("backend_mutated")
    return DeploymentHistory(
        id=row.id,
        environment_id=row.environment_id,
        environment_name=names.get(row.environment_id),
        binding_id=row.runtime_binding_id,
        target=row.target,
        status=row.status,
        created_at=_utc(row.created_at),
        verification_level=projected_level,
        independent_result=independent,
        backend_mutated=changed if isinstance(changed, bool) else None,
        error_digest=_digest(_dict(verification.get("apply_failed")).get("error_digest"))
        or _digest(_dict(row.receipt).get("error_digest")),
    )


@router.get("/api/v1/change-requests/{cr_id}/execution", response_model=ChangeExecution)
def get_execution(
    cr_id: str,
    response: Response,
    expanded: bool = False,
    session: Session = Depends(get_session),
    identity: Identity = Depends(get_identity),
):
    cr = session.scalar(
        select(ChangeRequest).where(ChangeRequest.id == cr_id, ChangeRequest.tenant_id == identity.tenant_id)
    )
    if cr is None:
        raise HTTPException(404, "not_found")
    ensure_permission(identity, "policy:read")
    dep_limit, audit_limit = (100, 200) if expanded else (20, 50)
    own_deployments = select(Deployment.id).where(
        Deployment.tenant_id == identity.tenant_id, Deployment.change_request_id == cr.id
    )
    rows = list(
        session.scalars(
            select(Deployment)
            .where(Deployment.id.in_(own_deployments))
            .order_by(Deployment.created_at.desc(), Deployment.id)
            .limit(dep_limit + 1)
        )
    )
    names = {}
    if identity.has_permission("env:read") and rows:
        names = dict(
            session.execute(
                select(Environment.id, Environment.name).where(
                    Environment.tenant_id == identity.tenant_id,
                    Environment.id.in_([d.environment_id for d in rows[:dep_limit]]),
                )
            ).all()
        )
    events = []
    can_audit = identity.has_permission("audit:read")
    if can_audit:
        events = list(
            session.scalars(
                select(AuditEvent)
                .where(
                    AuditEvent.tenant_id == identity.tenant_id,
                    or_(
                        and_(AuditEvent.resource_type == "change_request", AuditEvent.resource_id == cr.id),
                        and_(AuditEvent.resource_type == "deployment", AuditEvent.resource_id.in_(own_deployments)),
                    ),
                )
                .order_by(AuditEvent.created_at.desc(), AuditEvent.id)
                .limit(audit_limit + 1)
            )
        )
    response.headers["Cache-Control"] = "no-store"
    return ChangeExecution(
        change_id=cr.id,
        change_status=cr.status,
        evaluated_at=datetime.now(UTC),
        expanded=expanded,
        deployments=[_deployment(d, names) for d in rows[:dep_limit]],
        deployments_truncated=len(rows) > dep_limit,
        audit_access="allowed" if can_audit else "denied",
        audit_truncated=len(events) > audit_limit,
        audit_events=[
            AuditHistory(
                id=e.id,
                action=e.action,
                actor_id=e.actor_id,
                resource_id=e.resource_id,
                created_at=_utc(e.created_at),
                review_digest=_digest(_dict(e.summary).get("review_digest")),
                error_digest=_digest(_dict(e.summary).get("error_digest")),
            )
            for e in events[:audit_limit]
        ],
    )

File: apps/web/src/api/changeExecution.ts
import { get } from './client';

export interface DeploymentHistory {
  id: string; environment_id: string; environment_name: string | null; binding_id: string | null; target: string;
  status: string; created_at: string;
  verification_level: 'none' | 'config_readback' | 'behavior_enforced' | 'failed' | 'stale' | 'unknown';
  independent_result: 'not_checked' | 'verified' | 'mismatch' | 'unreachable' | 'no_receipt' | 'unknown';
  backend_mutated: boolean | null; error_digest: string | null;
}
export interface ChangeExecution {
  schema_version: 'change-execution/v1'; change_id: string; change_status: string; evaluated_at: string;
  deployments: DeploymentHistory[]; deployments_truncated: boolean; expanded: boolean;
  audit_access: 'allowed' | 'denied'; audit_truncated: boolean;
  audit_events: { id: string; action: string; actor_id: string; resource_id: string; created_at: string; review_digest: string | null; error_digest: string | null }[];
}
const object = (v: unknown): v is Record<string, unknown> => !!v && typeof v === 'object' && !Array.isArray(v);
const text = (v: unknown, max = 64): v is string => typeof v === 'string' && v.length <= max;
const nullableText = (v: unknown, max = 64) => v === null || text(v, max);
const oneOf = (v: unknown, allowed: readonly string[]): v is string => typeof v === 'string' && allowed.includes(v);
// Date.parse 会把 2026-02-30 之类不存在的日期滚动到邻近日期：用 UTC 回读校验日历日期真实存在。
const date = (v: unknown) => {
  if (!text(v, 40) || !/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z$/.test(v)) return false;
  const time = Date.parse(v);
  if (!Number.isFinite(time)) return false;
  const d = new Date(time);
  const pad = (n: number, width: number) => String(n).padStart(width, '0');
  return v.slice(0, 10) === `${pad(d.getUTCFullYear(), 4)}-${pad(d.getUTCMonth() + 1, 2)}-${pad(d.getUTCDate(), 2)}`;
};
const digest = (v: unknown) => v === null || typeof v === 'string' && /^[a-f0-9]{64}$/.test(v);
const exact = (v: Record<string, unknown>, keys: string[]) => Object.keys(v).length === keys.length && keys.every(k => k in v);
export function parseChangeExecution(v: unknown, id: string): ChangeExecution {
  if (!object(v) || !exact(v, ['schema_version', 'change_id', 'change_status', 'evaluated_at', 'deployments', 'deployments_truncated', 'audit_access', 'audit_events', 'audit_truncated', 'expanded']) || v.schema_version !== 'change-execution/v1' || v.change_id !== id || !text(v.change_status, 32) || !date(v.evaluated_at) || typeof v.expanded !== 'boolean' || typeof v.deployments_truncated !== 'boolean' || typeof v.audit_truncated !== 'boolean' || !oneOf(v.audit_access, ['allowed', 'denied'])) throw new Error('执行记录响应不完整或对象不匹配');
  if (!Array.isArray(v.deployments) || v.deployments.length > (v.expanded ? 100 : 20) || !v.deployments.every(d => object(d) && exact(d, ['id', 'environment_id', 'environment_name', 'binding_id', 'target', 'status', 'created_at', 'verification_level', 'independent_result', 'backend_mutated', 'error_digest']) && text(d.id) && text(d.environment_id) && nullableText(d.environment_name, 128) && nullableText(d.binding_id) && text(d.target, 128) && text(d.status, 32) && date(d.created_at) && oneOf(d.verification_level, ['none', 'config_readback', 'behavior_enforced', 'failed', 'stale', 'unknown']) && oneOf(d.independent_result, ['not_checked', 'verified', 'mismatch', 'unreachable', 'no_receipt', 'unknown']) && (d.backend_mutated === null || typeof d.backend_mutated === 'boolean') && digest(d.error_digest))) throw new Error('部署记录响应不完整');
  if (!Array.isArray(v.audit_events) || v.audit_events.length > (v.expanded ? 200 : 50) || !v.audit_events.every(e => object(e) && exact(e, ['id', 'action', 'actor_id', 'resource_id', 'created_at', 'review_digest', 'error_digest']) && text(e.id) && text(e.action) && text(e.actor_id, 256) && text(e.resource_id) && date(e.created_at) && digest(e.review_digest) && digest(e.error_digest)) || v.audit_access === 'denied' && (v.audit_events.length || v.audit_truncated)) throw new Error('审计记录响应不完整');
  if (new Set(v.deployments.map(d => d.id)).size !== v.deployments.length || new Set(v.audit_events.map(e => e.id)).size !== v.audit_events.length) throw new Error('结果包含重复记录');
  return v as unknown as ChangeExecution;
}
export async function readChangeExecution(id: string, expanded = false) {
  return parseChangeExecution(await get<unknown>(`/change-requests/${encodeURIComponent(id)}/execution`, { query: { expanded } }), id);
}

export const deploymentStatus = (status: string) => ({ pending: '待执行', sent: '已创建下发任务', verifying: '验证中', effective: '后端标记已生效', failed: '部署失败', rolled_back: '已回滚' }[status] ?? '状态待核对');
export function executionEvidence(d: DeploymentHistory): { label: string; detail: string; tone: 'ok' | 'warn' | 'err' } {
  if (d.status === 'failed') return { label: '部署失败，需核对执行端', detail: d.backend_mutated ? '已有修改后端的证据，请核对当前配置与回滚结果。' : '失败不证明后端没有变化；请结合错误摘要和执行端检查。', tone: 'err' };
  if (d.status === 'rolled_back') return { label: '已记录回滚', detail: '此前的部署验证不代表当前状态，请核对回滚后的执行端。', tone: 'warn' };
  if (d.independent_result === 'mismatch') return { label: '独立读回与部署回执不一致', detail: '可能存在带外变更；旧验证不能作为当前生效依据，请联系运维核对。', tone: 'err' };
  if (d.independent_result === 'unreachable') return { label: '最近独立读回无法连接执行端', detail: '暂不能确认当前状态；恢复连接后再核对。', tone: 'warn' };
  if (d.independent_result === 'no_receipt') return { label: '缺少可核对的部署回执', detail: '当前证据不足，不能确认执行状态。', tone: 'warn' };
  if (d.independent_result === 'unknown') return { label: '独立读回结果待核对', detail: '不能依赖此前的成功记录推断当前状态。', tone: 'warn' };
  if (d.verification_level === 'stale' || d.verification_level === 'failed') return { label: d.verification_level === 'stale' ? '验证证据已过期' : '验证未通过', detail: '需要重新核对执行端。', tone: 'err' };
  if (d.status !== 'effective') return { label: '尚无生效验证', detail: '任务已创建或正在处理，不代表配置已经应用。', tone: 'warn' };
  if (d.verification_level === 'config_readback') return { label: '配置已读回，行为未验证', detail: '历史读回证明当时配置一致，尚未证明实际工具调用受到拦截。', tone: 'warn' };
  if (d.verification_level === 'behavior_enforced') return { label: '已有行为验证记录', detail: '仅代表该次验证及其覆盖范围，不是持续保护保证。', tone: 'ok' };
  return { label: '验证证据不足', detail: '仅有后端状态或版本核对，不能确认完整配置及行为。', tone: 'warn' };
}

export const auditAction = (action: string) => ({ 'change.request.create': '提出变更', 'change.approve': '批准变更', 'change.reject': '驳回变更', 'deployment.reserve': '保存部署请求', 'deployment.create': '创建部署任务', 'deployment.verify': '核对部署配置', 'deployment.fail': '部署失败', 'deployment.receipt_verify': '独立核对部署回执', 'deployment.rollback': '回滚部署', 'deployment.rollback_fail': '回滚失败' }[action] ?? '其他审计操作');
